sstables/sstable_set.cc

```diff
@@ -464,6 +470,105 @@ class time_series_sstable_set : public sstable_set_impl {
     }
 };
 
+// Reader of a sstable together with its `max_position` (see `sstable::max_position`).
+struct reader_and_positions {
+    flat_mutation_reader reader;
+    position_in_partition max_position;
+
+    reader_and_positions(flat_mutation_reader r, position_in_partition max_pos)
+        : reader(std::move(r)), max_position(std::move(max_pos)) {}
+};
+
+// Queue of readers of sstables in a time_series_sstable_set,
+// returning readers in order of the sstables' min_position()s.
+//
+// Skips sstables that don't pass the supplied filter.
+//
+// The readers are created lazily on-demand using the supplied factory function.
+class min_position_reader_queue {
+    using container_t = time_series_sstable_set::container_t;
+    using value_t = container_t::value_type;
+
+    schema_ptr _schema;
+    lw_shared_ptr<const container_t> _sstables;
+
+    // Iterates over sstables in order of min_position().
+    container_t::const_iterator _it;
+    const container_t::const_iterator _end;
+
+    position_in_partition::tri_compare _cmp;
+
+    std::function<flat_mutation_reader(sstable&)> _create_reader;
+    std::function<bool(const sstable&)> _filter;
+
+    flat_mutation_reader create_reader(sstable& sst) {
+        return _create_reader(sst);
+    }
+
+    bool filter(const sstable& sst) const {
+        return _filter(sst);
+    }
+
+public:
+    min_position_reader_queue(schema_ptr schema,
+            lw_shared_ptr<const time_series_sstable_set::container_t> sstables,
+            std::function<flat_mutation_reader(sstable&)> create_reader,
+            std::function<bool(const sstable&)> filter)
+        : _schema(std::move(schema))
+        , _sstables(std::move(sstables))
+        , _it(_sstables->begin())
+        , _end(_sstables->end())
+        , _cmp(*_schema)
+        , _create_reader(std::move(create_reader))
+        , _filter(std::move(filter)) {}
+
+    // Open sstable readers to all sstables with smallest min_position() from the set
+    // {S: filter(S) and prev_min_pos < S.min_position() <= bound}, where `prev_min_pos` is the min_position()
+    // of the sstables returned from last non-empty pop() or -infinity if no sstables were previously returned,
+    // and `filter` is the filtering function provided when creating the queue.
+    //
+    // Note that there may be multiple returned sstables (all with the same position) or none.
+    //
+    // Note that S.min_position() is global for sstable S; if the readers are used to inspect specific partitions,
+    // the minimal positions in these partitions might actually all be greater than S.min_position().
+    std::vector<reader_and_positions> pop(position_in_partition_view bound, db::timeout_clock::time_point timeout) {
+        if (empty(bound)) {
+            return {};
+        }
+
+
+        auto it = std::find_if(_it, _end, [this, &bound] (const value_t& v) { return filter(*v.second); });
```

Comment:
`&bound` is unused.
Also, I don't understand what this does, it just finds the first sstable that passes the filter.

Reply:
So you do understand what this does.

Reply:
But I don't understand how it is correct given what we are trying to do here. This filter thing is muddying the waters.

Reply:
Well, we are interested only in sstables that pass the filter:
```
{S: filter(S) and prev_min_pos < S.min_position() <= bound}
```
(as the comment explains)
so I'm simply skipping sstables that do not.

Reply:
I think I understand now. I just noticed that we start from `_it` which I presume is at the current pos, so it is enough to just skip any sstables the filter rejects and we are at the right pos. The link is broken so I cannot go back to look at the code and verify.

Reply:
https://github.com/scylladb/scylla/pull/7437/commits/0445b3fcda14e4fad1004c4dd093345de09da927#diff-79da4c6195a1867e40a679218fb17f5e308e433983d351da9adbdcf32271e341R530

Reply:
Ok, looks correct.